### PR TITLE
[FLINK-28915] Support fetching artifacts in native K8s and standalone application mode

### PR DESCRIPTION
**Pull request by ferenc-csaky**
## What is the purpose of the change

Support fetching remote job JAR and additional artifacts (UDFs, formats, dependencies, etc.) in nativa Kubernetes and standalone application mode. The current change contains fetchers for DFS (via Flink FS abstraction) and HTTP. Builds on prior work from #20779 and addresses to comments on said PR.

## Brief change log

* In standalone app mode, a `--jars` option is added, which has unlimited args, which will be fetched before the Flink cluster start. Example:
```sh
./bin/standalone-job.sh start-foreground \
  --jars local:///path/to/flink-sandbox.jar,local:///path/to/test-udf.jar \
  --job-classname org.apache.flink.DummyJob
```
* In native K8s app mode, the user can define additional artifacts via the `user.artifacts.artifact-list` property. Example:
```sh
./bin/flink run-application \
  --target kubernetes-application \
  -Dkubernetes.cluster-id=flink-cluster \
  -Dkubernetes.container.image.ref=flink \
  -Duser.artifacts.raw-http-enabled=true \
  -Duser.artifacts.artifact-list=http://host.minikube.internal:9999/test-udf.jar \
  http://host.minikube.internal:9999/flink-sandbox.jar
```

## Verifying this change

This change added tests and can be verified as follows:

* Added tests for artifact fetching utils.
* Added tests for artifact fetching logic.
* Added tests to cover the changes in `DefaultPackagedProgramRetriever`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: Artifact deployment in native Kubernetes app mode.
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: docs/content.zh/docs/deployment/config.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/config.md:321]: Artifact fetching?

File: docs/content.zh/docs/deployment/config.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/config.md:323]: Flink can fetch user artifacts stored on remote DFS or accessible via an HTTP(S) endpoint.

File: docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md:110]: can we avoid baking in the version `1.17` into this cmd?

File: docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md:128]: like [Example] -> as described in this [example].

File: docs/content.zh/docs/deployment/resource-providers/standalone/docker.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/resource-providers/standalone/docker.md:124]: or pass jar path by *jar-file* argument -> or pass a list of jars via the `--jars` argument

File: docs/content.zh/docs/deployment/resource-providers/standalone/docker.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/resource-providers/standalone/docker.md:178]: `jars`

File: docs/content.zh/docs/deployment/resource-providers/standalone/docker.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/resource-providers/standalone/docker.md:329]: `--jars`

File: docs/content.zh/docs/deployment/resource-providers/standalone/kubernetes.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/resource-providers/standalone/kubernetes.md:123]: `--jars`

**Review by mbalassi — COMMENTED** [review]
Thanks @ferenc-csaky, I like this approach to the feature and thanks for preserving the prior work of @SwimSweet in separate commits (I might squash his 2 into a single commit later). On cursory look could you please update the docs too? I will review further later.

**Comment by ferenc-csaky** [comment]
> Thanks @ferenc-csaky, I like this approach to the feature and thanks for preserving the prior work of @SwimSweet in separate commits (I might squash his 2 into a single commit later). On cursory look could you please update the docs too? I will review further later.

Thank you for the review @mbalassi! Yes, I'll update the docs, furthermore I would like to highlight one thing regarding the current implementation that might be discussed a bit further:

Currently, for standalone app mode, the `--jars` arg handles both the job JAR and any additional artifact, which is a nice approach IMO, although it has some downsides:
1. The argument number is unlimited (as a job can have N additional dependency), so if someone puts the `--jars` CLI option last and tries to pass positional args afterwards all of those will be handled as "jars", for example `... --jars myjob.jar posarg1 posarg2`. This behavior could cause some confusion.
2. Because the additional artifacts are not differentiated from the job JAR, the fetching logic gets a bit more complicated.

I started thinking about using the newly introduced option as a dynamic property for standalone mode as well. WDYT?

**Comment by mbalassi** [comment]
@ferenc-csaky Thanks for the clarification. To your concern for accidentally passing additional arguments as jars, we have to be very careful as the user jars themselves often need arguments, this can have unintended consequences. I would like to ask you to instead expect the `--jars` argument as a comma separated list. This is exactly what Spark [does](https://spark.apache.org/docs/latest/submitting-applications.html#advanced-dependency-management).

File: flink-clients/src/main/java/org/apache/flink/client/program/artifact/ArtifactFetchManager.java
Comment (ferenc-csaky) [reply flink-clients/src/main/java/org/apache/flink/client/program/artifact/ArtifactFetchManager.java:153]: For an enterprise grade sw/usage, I think it makes sense to disable raw HTTP by default.

File: flink-clients/src/main/java/org/apache/flink/client/program/artifact/ArtifactFetchManager.java
Comment (ferenc-csaky) [reply flink-clients/src/main/java/org/apache/flink/client/program/artifact/ArtifactFetchManager.java:182]: Correct! Although after I thought this through again, this highlights a bug in the current behavior, because the `LocalArtifactFetcher` at the moment does not copies the artifacts, only gets their local path, and in case we pass for example `local:///tmp/foo/foo.jar, local:///tmp/baz/baz.jar` only 1 of them will be loaded to the CP, cause the `DefaultPackagedProgramRetriever` won't know about the other dir.

2 possible solutions that come into my mind would be:
1. Do copy `local://` artifacts to the `ARTIFACT_BASE_DIR` as well.
2. Instead of passing the `ARTIFACT_BASE_DIR`, pass the list of artifacts to `DefaultPackagedProgramRetriever`.

I do not really like option 1. because of doubling the required disk space. Option 2 should not add much complexity anyways, I went with passing the directory because this way it was possible to reuse loading the contents as it almost the same as `userLibDir`.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesApplicationClusterEntrypoint.java
Comment (ferenc-csaky) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesApplicationClusterEntrypoint.java]: I did not write this part and personally have limited K8s experience, but I think the reason behind this is that multiple deployments may end up on the same node, so adding the `namespace` and `cluster-id` to the dir path makes sure any deployment will have its own artifact base dir.

**Comment by ferenc-csaky** [comment]
@flinkbot run azure

File: docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md:122]: nit: remove `Now,`

File: docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md:127]: nit: `downloaded by`

File: docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md
Comment (mbalassi) [inline docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md:349]: nit: `JobManager pod's`

File: docs/layouts/shortcodes/generated/artifact_fetch_configuration.html
Comment (mbalassi) [inline docs/layouts/shortcodes/generated/artifact_fetch_configuration.html:15]: Why are we switching to semicolon seperator from commas as used by `--jars`?

**Review by mbalassi — COMMENTED** [review]
Looks good, thanks for the extensive testing. I left some more minor comments, ready to approve and merge it once those are addressed.

File: docs/layouts/shortcodes/generated/artifact_fetch_configuration.html
Comment (ferenc-csaky) [reply docs/layouts/shortcodes/generated/artifact_fetch_configuration.html:15]: This is the behavior of the [`ConfigurationUtils`](https://github.com/apache/flink/blob/2cf213ec9e4db81815a9b37013904aa47a94aa01/flink-core/src/main/java/org/apache/flink/configuration/ConfigurationUtils.java#L439), I wrongly assumed before that the separator is a comma, but it has to be a semicolon. Any other `List<String>` typed `ConfigOption` works the same, e.g. the ones in [`CoreOptions`](https://github.com/apache/flink/blob/2cf213ec9e4db81815a9b37013904aa47a94aa01/flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java#L107).

**Review by mbalassi — APPROVED** [review]
Thanks, lgtm. I am squshing the first 2 commits manually before merging.

File: docs/layouts/shortcodes/generated/artifact_fetch_configuration.html
Comment (mbalassi) [reply docs/layouts/shortcodes/generated/artifact_fetch_configuration.html:15]: Thanks for the clarification. 😐

**Pull request closed without merging**
